media-libs/openimageio: add 2.4.12.0

negril opened the pull request:
moved `qt5` and `opengl` use-flag into `iv` use-flag mirroring the logic in `src/cmake/externalpackages.cmake:252`

added qt6 support

AndrewAmmerlaan commented on media-libs/openimageio/openimageio-2.4.12.0.ebuild:
```diff
@@ -0,0 +1,187 @@
+		virtual/glu
+		virtual/opengl
+		|| (
+			dev-qt/qtbase:6[gui,widgets,opengl]
```
As per qt policy we are going to need to add a `qt6` flag for toggling this: https://wiki.gentoo.org/wiki/Project:Qt/Policies#Handling_different_versions_of_Qt

AndrewAmmerlaan replied:
Also there needs to be some switch in the configure phase to actually instruct the build system to use qt5 or qt6. 

The user can have both installed at the same time and we can't let the build system automagically decide which to use.

negril replied:
From https://github.com/OpenImageIO/oiio/blob/v2.4.12.0/INSTALL.md?plain=1#L33 :
```md
### Optional dependencies -- features may be disabled if not found
* If you are building the iv viewer (which will be disabled if any of these are not found):
     * Qt5 >= 5.6 (tested through 5.15) or Qt6 (tested through 6.4)
     * OpenGL
```

https://github.com/OpenImageIO/oiio/blob/v2.4.12.0/src/cmake/externalpackages.cmake#L252:
```cmake
# Qt -- used for iv
option (USE_QT "Use Qt if found" ON)
if (USE_QT)
    checked_find_package (OpenGL)   # used for iv
endif ()
if (USE_QT AND OPENGL_FOUND)
    checked_find_package (Qt6 COMPONENTS Core Gui Widgets OpenGLWidgets)
    if (NOT Qt6_FOUND)
        checked_find_package (Qt5 COMPONENTS Core Gui Widgets OpenGL)
    endif ()
    if (NOT Qt5_FOUND AND NOT Qt6_FOUND AND APPLE)
        message (STATUS "  If you think you installed qt with Homebrew and it still doesn't work,")
        message (STATUS "  try:   export PATH=/usr/local/opt/qt/bin:$PATH")
    endif ()
endif ()
´´´
So OpenGL is required for and only for `iv`.

It wouldn't care if you set `qt5` or `qt6`. If you know a way of making it adhere to that let me know.

There are `USE_OPENGL` and `USE_QT` but that is only relevant for `iv`. So you could leave the `qt5`, `qt6` and `opengl` flags in, but that is just a very convoluted way of saying build `iv`.
And to make sure `qt5` builds against Qt5 you would have to add blockers against `dev-qt/qtbase:6`, I am not sure if that is good design, that's why I ended up with this.

AndrewAmmerlaan replied:
`cmake_use_find_package` from `cmake.eclass` can be used to prevent the build system from finding an installed package if the use flag is not enabled.

https://devmanual.gentoo.org/eclass-reference/cmake.eclass/index.html

negril replied:
That seems like way to much work, to fix the qt detection that should be fixed upstream and would have gone unnoticed had I not looked into the source that much. Considering this is done to help fix the fact that this currently does not build due to:
```
CMake Error at /usr/lib64/cmake/OpenColorIO/OpenColorIOTargets.cmake:70 (set_target_properties):
  The link interface of target "OpenColorIO::OpenColorIO" contains:
 
    expat::expat
 
  but the target was not found.  Possible reasons include:
 
    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
 
Call Stack (most recent call first):
  /usr/lib64/cmake/OpenColorIO/OpenColorIOConfig.cmake:22 (include)
  src/cmake/modules/FindOpenColorIO.cmake:31 (find_package)
  src/cmake/checked_find_package.cmake:127 (find_package)
  src/cmake/externalpackages.cmake:177 (checked_find_package)
  CMakeLists.txt:155 (include)
```

So we could either mask the `iv` USE-flag or I add:
```sh
USE=+="qt5 qt6"
REQUIRED_USE="^^ ( qt5 qt6 )"
BDEPEND+="qt6? ( !dev-qt/qtbase:6 )"
```
seeing that `iv` is a tool not the main feature for `openimageio`.

AndrewAmmerlaan reviewed (changes requested):
Instead of the `iv` flag, I would use `qt5` and `qt6` flags. 

OpenGL support is no longer optional when building the gui?

AndrewAmmerlaan commented on media-libs/openimageio/openimageio-2.4.12.0.ebuild:
```diff
@@ -0,0 +1,187 @@
+		-DUSE_PYTHON=$(usex python)
+		-DUSE_LIBRAW=$(usex raw)
+		-DUSE_FREETYPE=$(usex truetype)
+		-DUSE_SIMD=$(local IFS=','; echo "${mysimd[*]}")
```
```suggestion
		-DUSE_SIMD=$(local IFS=','; echo "${mysimd[*]}")
		"$(cmake_use_find_package qt5 Qt5)"
		"$(cmake_use_find_package qt6 Qt6)"
```

Actually it is not a lot of work, this should do the trick. You are correct though that this should really be fixed upstream.

Blockers is not an option, this is just going to create an unpleasant mess for users.

AndrewAmmerlaan commented on media-libs/openimageio/openimageio-2.4.12.0.ebuild:
```diff
@@ -0,0 +1,186 @@
+	dicom? ( sci-libs/dcmtk )
+	ffmpeg? ( media-video/ffmpeg:= )
+	gif? ( media-libs/giflib:0= )
+	iv? (
+		media-libs/glew:=
+		virtual/glu
+		virtual/opengl
+		qt5? (
+			dev-qt/qtcore:5
+			dev-qt/qtgui:5
+			dev-qt/qtopengl:5
+			dev-qt/qtwidgets:5
+		)
+		qt6? ( dev-qt/qtbase:6[gui,widgets,opengl] )
```
```suggestion
	qt5? (
		media-libs/glew:=
		virtual/glu
		virtual/opengl
		dev-qt/qtcore:5
		dev-qt/qtgui:5
		dev-qt/qtopengl:5
		dev-qt/qtwidgets:5
	)
	qt6? ( 
		media-libs/glew:=
		virtual/glu
		dev-qt/qtbase:6[gui,widgets,opengl] 
	)
```

negril commented on media-libs/openimageio/metadata.xml:
```diff
@@ -7,9 +7,12 @@
 	</maintainer>
 	<use>
 		<flag name="dicom">Enable DICOM medical image file support via <pkg>sci-libs/dcmtk</pkg></flag>
+		<flag name="iv">Enable iv (image viewer) tool</flag>
```
Are you okay with the `tools` useflag? Building the tools wasn't optional before.

My thought was to make it more obvious that the `qt{5,6}` flags builds only iv. But I don't care either way, so I will revert `iv` to `qt5`.

AndrewAmmerlaan replied:
> Are you okay with the `tools` useflag? Building the tools wasn't optional before.

I think `tools` is fine since it is enabled by default. So introducing that flag shouldn't cause any surprises on upgrade.

> My thought was to make it more obvious that the qt{5,6} flags builds only iv. But I don't care either way, so I will revert iv to qt5.

`iv` does indeed make it more obvious what the flag does. However, I don't think the increase in clarity ways up against the cost of end users having to reconfigure and re-emerge. 

Unfortunately sometimes we are stuck with poor use flag names because the principle of least surprise tells us that changing the name of the flag is a bad idea. For the renaming of packages we have mechanisms in place to make this possible without inconveniencing the end user too much, there is currently no such mechanism for use flags unfortunately.

AndrewAmmerlaan commented on media-libs/openimageio/openimageio-2.4.12.0.ebuild:
```diff
@@ -0,0 +1,185 @@
+	)
+
+	use qt5 && mycmakeargs+=( $(cmake_use_find_package qt5 Qt5) )
+	use qt6 && mycmakeargs+=( $(cmake_use_find_package qt6 Qt6) )
```
This way it still won't work :(
`cmake_use_find_package qt5 Qt5` echo's `-DCMAKE_DISABLE_FIND_PACKAGE_Qt5=OFF` if the flag is enabled and `-DCMAKE_DISABLE_FIND_PACKAGE_Qt5=ON` if it is disabled. Therefore this code does nothing, if qt5 is enabled it will _not_ disable finding qt5 (i.e. do nothing differently), and if qt6 is enabled it will _not_ disable finding qt6. What we want is to disable finding qt5 if we are using qt6, and to disable finding qt6 if we are using qt5. 

It would work if you remove the `use qt5 &&` stuff as in the example in my previous post. However, as per Qt QA policy we shouldn't do `REQUIRED_USE=?? (q5 qt6)`, instead what we want is to enable the image viewer if we have qt5 or qt6, and prefer qt5 unless we have the qt6 flag enabled. So:
```
   if use qt5 || use qt6; then
       mycmakeargs+=( -DENABLE_IV=ON ) 
       if use qt6; then
           mycmakeargs+=( -DCMAKE_DISABLE_FIND_PACKAGE_Qt5=ON ) 
       else
           mycmakeargs+=( -DCMAKE_DISABLE_FIND_PACKAGE_Qt6=ON ) 
       fi
   else
       mycmakeargs+=( -DENABLE_IV=OFF ) 
   fi
```
Sorry, I should have realized the REQUIRED_USE problem before and adjusted the example in my previous post accordingly.

AndrewAmmerlaan commented:
Thanks for the PR and for your patience :+1: 

The opengl dependencies got lost somewhere along the way, since we definitely do need them I amended them in:
```
		media-libs/glew:=
		virtual/glu
		virtual/opengl
```

negril commented:
I dropped `media-libs/glew` since it's not longer used outside of `Nuke` and hasn't been since version 2.0. See https://github.com/OpenImageIO/oiio/blob/v2.4.12.0/CHANGES.md?plain=1#L4010.

Same goes for `virtual/glu` `virtual/opengl`. My `iv` links against `/usr/lib64/libOpenGL.so.0` which belongs to `media-libs/libglvnd`. Which among others is pulled in by `dev-qt/qtgui`. So I saw no reason to depend on either.

I am also vaguely remember that Qt uses GLES and not GLu. 

Either way I don't think we need these dependencies, hence I dropped them.

AndrewAmmerlaan commented:
> I dropped `media-libs/glew` since it's not longer used outside of `Nuke` and hasn't been since version 2.0. See https://github.com/OpenImageIO/oiio/blob/v2.4.12.0/CHANGES.md?plain=1#L4010.

Aaah right, I'll drop them later then.
 
> Same goes for `virtual/glu` `virtual/opengl`. My `iv` links against `/usr/lib64/libOpenGL.so.0` which belongs to `media-libs/libglvnd`. Which among others is pulled in by `dev-qt/qtgui`. So I saw no reason to depend on either.

We shouldn't rely on implicit dependencies though, always best to specify them explicitly. This helps also when finding reverse dependencies (i.e. finding all consumers of opengl on given system or in the tree).

thesamesam commented:
That said, it _is_ true we should depend on libglvnd instead of virtual/opengl usually, I think. But we haven't formally deprecated virtual/opengl yet.

See https://wiki.gentoo.org/wiki/User:ManDay/Deprecating_the_virtual_OpenGL_ebuild. But they got impatient, I think, and then didn't follow it through.

AndrewAmmerlaan commented:
> That said, it _is_ true we should depend on libglvnd instead of virtual/opengl usually, I think. But we haven't formally deprecated virtual/opengl yet.

The virtual depends on mesa which depends on libglvnd, but it also `|| ()` depends on the mingw runtime. So depending on the virtual versus depending on libglvnd directly gives a different result (if mingw runtime is installed). What is the deal with that? Can we use the mingw runtime as an alternative opengl provider?

```
RDEPEND="
	|| (
		>=media-libs/mesa-9.1.6[X(+),${MULTILIB_USEDEP}]
		dev-util/mingw64-runtime
	)"
```

thesamesam commented:
You'd have to ask @ionenwks.

negril commented:
`iv` links against `libOpenGL.so.0` for me, so reading [this](https://wiki.gentoo.org/wiki/User:ManDay/Deprecating_the_virtual_OpenGL_ebuild) depending on `media-libs/libglvnd` would be the solution?

And then someone should make an effort to get rid of the consumers of `virtual/opengl`.

We also need to fix https://bugs.gentoo.org/909280.

AndrewAmmerlaan commented:
Thanks for clarifying @ionenwks
